torch/csrc/jit/dynamic_dag.h

```diff
@@ -0,0 +1,388 @@
+#pragma once
+
+#include <algorithm>
+#include <iostream>
+#include <unordered_set>
+#include <vector>
+
+#include "torch/csrc/utils/functional.h"
+
+namespace torch { namespace jit { namespace detail {
+
+// DynamicDAG is a DAG that dynamically maintains a topological order as
+// edges/vertices are added and removed.
+//
+// The implementation is based off of the following paper:
+// "A Dynamic Topsort Algorithm for Directed Acyclic Graphs"
+// by David Pearce and Paul Kelly
+// https://www.doc.ic.ac.uk/~phjk/Publications/DynamicTopoSortAlg-JEA-07.pdf
+
+template <typename T> struct Vertex;
+
+template <typename T>
+using vertex_set = std::unordered_set<Vertex<T>*>;
+
+template <typename T>
+using vertex_list = std::vector<Vertex<T>*>;
+
+template <typename T>
+using maybe_vertex_list = std::vector<at::optional<Vertex<T>*>>;
+
+template <typename T>
+struct Vertex {
+  std::vector<T> rdata; // stored in reverse order
+  size_t ord; // index in topological ordering. Is unique.
+  bool visited; // If this vertex has been visited
+  vertex_set<T> in_edges;
+  vertex_set<T> out_edges;
+
+  std::string toString();
+};
+
+template <typename T>
+struct DynamicDAG {
+  ~DynamicDAG();
+
+  Vertex<T>* newVertex(T datum);
+  std::pair<vertex_set<T>,vertex_set<T>> removeVertex(Vertex<T>* v);
```

Comment:
Can we please make a helper struct for this (think of it as a named tuple). Otherwise it's easy to mix up those arguments.